program-runtime/src/sysvar_cache.rs

```diff
@@ -23,92 +24,213 @@ impl ::solana_frozen_abi::abi_example::AbiExample for SysvarCache {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct SysvarWithLamports<T: Sysvar> {
+    value: Arc<T>,
+    lamports: u64,
+}
+impl<T: Sysvar> SysvarWithLamports<T> {
+    pub fn new(value: T, lamports: u64) -> Self {
+        Self {
+            value: Arc::new(value),
+            lamports,
+        }
+    }
+    pub fn into_account(&self, rent_epoch: Epoch) -> Account {
+        create_account_with_fields(&*self.value, (self.lamports, rent_epoch))
+    }
+}
+impl<'a, T: Sysvar> TryFrom<&'a AccountSharedData> for SysvarWithLamports<T> {
+    type Error = InstructionError;
+    fn try_from(account: &'a AccountSharedData) -> Result<Self, Self::Error> {
+        let value = bincode::deserialize(account.data())
+            .map_err(|_| InstructionError::UnsupportedSysvar)?;
+        Ok(SysvarWithLamports {
+            value,
+            lamports: account.lamports(),
+        })
+    }
+}
+
 #[derive(Default, Clone, Debug)]
 pub struct SysvarCache {
-    clock: Option<Arc<Clock>>,
-    epoch_schedule: Option<Arc<EpochSchedule>>,
+    /// Used to simulate the rent_epoch on real accounts
+    rent_epoch: Epoch,
+    clock: Option<SysvarWithLamports<Clock>>,
+    epoch_schedule: Option<SysvarWithLamports<EpochSchedule>>,
     #[allow(deprecated)]
-    fees: Option<Arc<Fees>>,
-    rent: Option<Arc<Rent>>,
-    slot_hashes: Option<Arc<SlotHashes>>,
+    fees: Option<SysvarWithLamports<Fees>>,
+    rent: Option<SysvarWithLamports<Rent>>,
+    slot_hashes: Option<SysvarWithLamports<SlotHashes>>,
     #[allow(deprecated)]
-    recent_blockhashes: Option<Arc<RecentBlockhashes>>,
-    stake_history: Option<Arc<StakeHistory>>,
+    recent_blockhashes: Option<SysvarWithLamports<RecentBlockhashes>>,
+    stake_history: Option<SysvarWithLamports<StakeHistory>>,
 }
 
 impl SysvarCache {
+    pub fn new(rent_epoch: Epoch) -> Self {
+        Self {
+            rent_epoch,
+            ..Self::default()
+        }
+    }
+
+    pub fn get_account(&self, pubkey: &Pubkey) -> Result<AccountSharedData, InstructionError> {
```

Comment:
Good point, it would avoid a lot of potential synchronization issues too.  Just to confirm, that would mean storing both the deserialized type and the serialized account.

Reply:
Yes, exactly